package_modules/yum

```diff
@@ -0,0 +1,190 @@
+#!/usr/bin/python
+
+import sys
+import os
+import subprocess
+import re
+
+
+rpm_cmd = os.environ.get('CFENGINE_TEST_RPM_CMD', "/bin/rpm")
+rpm_output_format = "Name=%{name}\nVersion=%{version}\nArchitecture=%{arch}\n"
+
+yum_cmd = os.environ.get('CFENGINE_TEST_YUM_CMD', "/usr/bin/yum")
+yum_options = "--quiet"
+
+def get_package_data():
+    pkg_string = ""
+    for line in sys.stdin:
+        if line.startswith("File="):
+            pkg_string = line.split("=", 1)[1].rstrip()
+            break
+
+    if not pkg_string:
+        return 1
+
+    repo = False
+    if pkg_string.startswith("/"):
+        # Absolute file.
+        sys.stdout.write("PackageType=file\n")
+        sys.stdout.flush()
+        return subprocess.call([rpm_cmd, "--qf", rpm_output_format, "-qp", result.group(1)])
+    elif re.search("([:,]|-[0-9])", pkg_string):
+        # Contains either a version number or an illegal symbol.
+        sys.stdout.write(line + "ErrorMessage: Package string with illegal format\n")
+        return 1
+    else:
+        repo = True
+        sys.stdout.write("PackageType=repo\n")
+        sys.stdout.write("Name=" + pkg_string + "\n")
+        return 0
+
+
+def list_installed():
+    # Ignore everything.
+    sys.stdin.readlines()
+
+    return subprocess.call([rpm_cmd, "-qa", "--qf", rpm_output_format])
+
+
+def list_updates(online):
+    # Ignore everything.
+    sys.stdin.readlines()
+
+    online_flag = [] if online else ["-C"]
+
+    process = subprocess.Popen([yum_cmd, yum_options] + online_flag + ["check-update"], stdout=subprocess.PIPE)
+    lastline = ""
+    for line in process.stdout:
+        # Combine multiline entries into one line. A line without at least three
+        # space separated fields gets combined with the next line, if that line
+        # starts with a space.
+        if lastline and not line[0].isspace():
+            # Line does not start with a space. No combination.
+            lastline = ""
+
+        line = lastline + line
+        match = re.match("^\S+\s+\S+\s+\S+", line)
+        if match == None:
+            # Keep line but strip trailing newline.
+            lastline = line[:-1]
+            continue
+
+        lastline = ""
+        match = re.match("^(?P<name>\S+)\.(?P<arch>[^.\s]+)\s+(?P<version>\S+)\s+\S+\s*$", line)
+        if match != None:
+            sys.stdout.write("Name=" + match.group("name") + "\n")
+            sys.stdout.write("Version=" + match.group("version") + "\n")
+            sys.stdout.write("Architecture=" + match.group("arch") + "\n")
```

Comment:
Concatenating many lines into one, while repeatedly matching a "three words" regex, is more expensive than accumulating the results of line.split() in a words list until its length is three (or more, in which case discard it).

Another way to write this code, assuming you prefer to join the continuation lines, would be to define a generator first,
```
def joinedlines(src):
    src = iter(src) # get the iterator over it, in case it's not already a generator
    joined = src.next()
    for line in src:
        if line[0].isspace(): joined += line
        else:
            yield joined
            joined = line

    if joined: yield joined
```
Because this "function" uses yield, it's automagically a generator and raises StopIteration when it falls off the end.
You can then just iterate `joinedlines(process.stdout)` applying your second regex to each line, without bothering with the first regex check at all.
I'm not sure what the exact semantics of your continuation lines is, though: I've assumed that if a fifty-word line is split up with a space at the start of each continuation, it should be counted as one line.  If it isn't, your code groups up its first three words (and uses them) then groups up the remaining lines into a single long line (the first of which starts with a space so doesn't match either regex) until you hit the next line to start with a non-space, at which point you'll discard this long line and start building a new one.  This leads to whether or not you match depending on the pattern of continuation lines; if I give you
```
one
    two
    three
    four
    ...
    fifty
```
you'll match the initial `one two three` and discard the rest; but if I give you
```
one
    two
    three four
    ...
    fifty
```
you won't match anything; you'll initially build up a line until it has the first four words, then clear lastline and fail to match the line (too many words); after that, you'll accumulate the rest in lastline (starting with a space, so matching neither pattern) until you hit the next line that doesn't start with a blank, when you'll throw away the whole lot and start over.